constraint-solver/src/indexed_constraint_system.rs

```diff
@@ -90,34 +93,63 @@ impl<T, V> Default for IndexedConstraintSystemWithQueue<T, V> {
     }
 }
 
-/// A reference to an item in the constraint system.
+/// A reference to an item in the constraint system, based on the index.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
 enum ConstraintSystemItem {
+    /// A reference to an algebraic constraint.
     AlgebraicConstraint(usize),
+    /// A reference to a bus interaction.
     BusInteraction(usize),
+    /// A reference to a derived variable. This is only used internal to the
+    /// IndexedConstraintSystem.
+    DerivedVariable(usize),
 }
 
 impl ConstraintSystemItem {
     /// Returns an index that is unique across both algebraic constraints and bus interactions.
-    fn flat_id(&self) -> usize {
+    /// Panics for derived variables.
+    fn flat_constraint_id(&self) -> usize {
         match self {
             ConstraintSystemItem::AlgebraicConstraint(i) => 2 * i,
             ConstraintSystemItem::BusInteraction(i) => 2 * i + 1,
+            ConstraintSystemItem::DerivedVariable(_) => panic!(),
         }
     }
 
+    /// Returns the index of the item. Note that the indices are not disjoint between different kinds
+    /// of items.
+    fn index(&self) -> usize {
+        match self {
+            ConstraintSystemItem::AlgebraicConstraint(index)
+            | ConstraintSystemItem::BusInteraction(index)
+            | ConstraintSystemItem::DerivedVariable(index) => *index,
+        }
+    }
+
+    /// Returns true if this constraint system item is an actual constraint and not a derived variable.
```

Comment:
`is_constraint` sounds like algebraic constraint to me. Maybe we can use `!is_derived_vairable`? No strong opinion though.